packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -88,40 +92,69 @@ class TimelineData {
 
   void _handleFrameStartEvent(TraceEvent event) {
     if (event.id != null) {
-      if (_pendingFrames[event.id] == null) {
+      final String id = _getFrameId(event);
+      if (_pendingFrames[id] == null) {
         // Create a new TimelineFrame if we do not already have one for this id.
-        _pendingFrames[event.id] = TimelineFrame(event.id);
+        _pendingFrames[id] = TimelineFrame(id);
       }
-      if (_pendingFrames[event.id].startTime == null) {
-        _pendingFrames[event.id].startTime = event.timestampMicros;
-        _maybeAddPendingEvents();
-      } else {
-        // TODO(kenzie): investigate and prevent this case.
-        print('Error - already set startTime '
-            '${_pendingFrames[event.id].startTime} for frame ${event.id}.\n'
-            'TraceEvent - ${event.json.toString()}');
+
+      // Drop any duplicate trace events. Log a warning to console if the
+      // timestamp delta exceeds [duplicateTraceWarningThreshold].
+      if (_pendingFrames[id].startTime != null) {
+        final delta =
+            (_pendingFrames[id].startTime - event.timestampMicros).abs();
+        if (delta > duplicateTraceWarningThreshold) {
+          print('Warning - received duplicate FrameStart event for frame $id.\n'
+              'Attempting to overwrite startTime '
+              '${_pendingFrames[id].startTime} with new value: '
+              '${event.timestampMicros}.\n Delta = $delta.\n If you see this'
+              'warning, please copy and paste this message into a new issue at '
+              'https://github.com/flutter/devtools/issues/new, and CC '
+              '@kenzieschmoll.');
+        }
+        return;
       }
+
+      _pendingFrames[id].startTime = event.timestampMicros;
```

Comment:
shouldn't we make the startTime for the frame just be the minimum of all timestamps for the frame?
Same for end time.

Reply:
We still need these start and end times from the flow events to determine the window for the frame. Then once we have both start and end time, we'll try to fit cpu and gpu chunks within that window (using an epsilon to determine "fit"). The getters for both frame startTime and endTime take into account the cpuStartTime and and gpuEndTime respectively, so the flame chart reflects the absolute start/end times of events.

Reply:
Discussed offline. Removed warning. Will take min start time and max end time to be the truth.